pkg/kv/kvserver/loqrecovery/store.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package loqrecovery
+
+import (
+	"io"
+	"io/fs"
+	"regexp"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/loqrecovery/loqrecoverypb"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/pebble/vfs"
+)
+
+const recoveryPlanDir = "loq-recovery-plans"
+const extension = ".json"
+
+type PlanStore struct {
+	path string
+	fs   vfs.FS
+}
+
+type StoredPlan struct {
+	PlanID       uuid.UUID
+	CreationTime time.Time
+}
+
+func (p StoredPlan) Empty() bool {
+	return p.PlanID.Equal(uuid.UUID{})
+}
+
+func (p StoredPlan) SamePlan(planID uuid.UUID) bool {
+	return planID.Equal(p.PlanID)
+}
+
+// NewPlanStore creates a plan store for recovery.
+func NewPlanStore(path string, fs vfs.FS) PlanStore {
+	return PlanStore{
+		path: fs.PathJoin(path, recoveryPlanDir),
+		fs:   fs,
+	}
+}
+
+func (s PlanStore) HasPlan() (StoredPlan, error) {
+	infos, err := s.listPlanFiles()
+	if err != nil {
+		return StoredPlan{}, err
+	}
+	if len(infos) > 1 {
+		return StoredPlan{}, errors.Newf("more than a single recovery plan exists in store: %s, %s",
+			s.fs.PathJoin(s.path, infos[0].Name()),
+			s.fs.PathJoin(s.path, infos[1].Name()))
+	}
+	if len(infos) > 0 {
+		name := infos[0].Name()
+		planID, err := uuid.FromString(name[:len(name)-len(extension)])
+		if err != nil {
+			return StoredPlan{}, errors.Newf("found corrupted file name in loq recovery store: %s",
+				s.fs.PathJoin(s.path, infos[0].Name()))
+		}
+		return StoredPlan{
+			PlanID:       planID,
+			CreationTime: infos[0].ModTime(),
+		}, nil
+	}
+	return StoredPlan{}, nil
+}
+
+func (s PlanStore) RemovePlans() error {
+	infos, err := s.listPlanFiles()
+	if err != nil {
+		return err
+	}
+	for _, i := range infos {
+		if err := s.fs.Remove(s.fs.PathJoin(s.path, i.Name())); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s PlanStore) SavePlan(plan loqrecoverypb.ReplicaUpdatePlan) error {
+	existing, err := s.HasPlan()
+	if err != nil {
+		return errors.Wrap(err, "failed to perform existing recovery plan check")
+	}
+	if existing.SamePlan(plan.PlanID) {
+		// Plan is already set, nothing to do.
+		return nil
+	}
+	if !existing.Empty() {
+		return errors.Newf("can not set recovery plan %s, plan %s is already staged", plan.PlanID,
```

Comment:
I believe we said previously that we can't "unstage" a plan because we've already tombstones the nodes. However, should it be possible to replace the plan, e.g. in case something was wrong with the initial plan?

Reply:
We can't unstage but we can't guarantee that new plan ends up on the same set of nodes. So individual nodes can technically have plan removed.